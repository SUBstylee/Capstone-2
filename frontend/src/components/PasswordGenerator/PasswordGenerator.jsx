import {useState} from 'react';
import './PasswordGenerator.scss';

const PasswordGenerator = () => {
    const [result,setResult]=useState('');
    const [length,setLength]=useState(10);
    const [upper,setUpper]=useState(true);
    const [lower,setLower]=useState(true);
    const [number,setNumber]=useState(true);
    const [symbol,setSymbol]=useState(true);

    const getRandLower=()=>{
        return String.fromCharCode(Math.floor(Math.random()*26)+97);
    };
    const getRandUpper=()=>{
        return String.fromCharCode(Math.floor(Math.random()*26)+65);
    };
    const getRandNumber=()=>{
        return String.fromCharCode(Math.floor(Math.random()*10)+48);
    };
    const getRandSymbol=()=>{
        const symbols='!@#$%^&*()+<>/,.';
        return symbols[Math.floor(Math.random()*symbols.length)];
    };
    const randomFunc={
        lower:getRandLower,
        upper:getRandUpper,
        number:getRandNumber,
        symbol:getRandSymbol
    }
    const generatePassword=(upper,lower,number,symbol,length)=>{
        let generatedPassword='';
        const typesCount=lower+upper+number+symbol;
        const typesArr=[{upper},{lower},{number},{symbol}].filter((item)=>Object.values(item)[0]);
        if(typesCount===0)return '';
        for (let i = 0; i < length; i++) {
            const funcName = Object.keys(typesArr[Math.floor(Math.random() * typesCount)])[0];
            generatedPassword += randomFunc[funcName]();
          }
          return generatedPassword;
    };
    const handleClick=()=>{
        setResult(generatePassword(upper,lower,number,symbol,length))
    };
    const clipboardClick=()=>{
        if(!result)return;
        const copied=navigator.clipboard.writeText(result);
        alert('Password copied to clipboard!');
        return copied;   
    };

    return (
        <div className='PasswordGenerator'>
            <h2>Password Generator</h2>
            <div className="resultContainer">
                <span id="result">{result}</span>
                <button className='btn' id='clipboard' onClick={clipboardClick}><i className="far fa-clipboard"></i></button>
            </div>
            <div className="settings">
                <div className="setting">
                    <label htmlFor='length'>Password Length (6-20)</label>
                    <input type='number' id='length' min='6' max='20' value={length} onChange={(e)=>setLength(e.target.value)}/>
                </div>
                <div className="setting">
                    <label htmlFor='uppercase'>Include uppercase letters</label>
                    <input type="checkbox" id='uppercase' checked={upper} onChange={(e)=>setUpper(!upper)}/>
                </div>
                <div className="setting">
                    <label htmlFor='lowercase'>Include lowercase letters</label>
                    <input type="checkbox" id='lowercase' checked={lower} onChange={(e)=>setLower(!lower)}/>
                </div>
                <div className="setting">
                    <label htmlFor='numbers'>Include numbers</label>
                    <input type="checkbox" id='numbers' checked={number} onChange={(e)=>setNumber(!number)}/>
                </div>
                <div className="setting">
                    <label htmlFor='symbols'>Include symbols</label>
                    <input type="checkbox" id='symbols' checked={symbol} onChange={(e)=>setSymbol(!symbol)}/>
                </div>
            </div>
            <button className='btn btn-large' id='generate' onClick={handleClick}>Generate Password</button>
        </div>
    )
}

export default PasswordGenerator