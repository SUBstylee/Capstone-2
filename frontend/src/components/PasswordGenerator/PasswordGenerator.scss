.PasswordGenerator{
    background-color: #eeeeee;
    box-shadow: 0px 2px 10px rgba($color: #505050, $alpha: 0.2);
    padding: 20px;
    width: 350px;
    border-radius: 8px;
    margin-top: 30px;

    h2{
        margin: 10px 0 20px;
        text-align: center;
    }
    .resultContainer{
        background-color: #fff;
        display: flex;
        justify-content: flex-start;
        align-items: center;
        position: relative;
        font-size: 18px;
        letter-spacing: 1px;
        padding: 12px 10px;
        height: 50px;
        width: 100%;
    }
    .resultContainer #result{
        word-wrap: break-word;
        max-width: calc(100% - 40px);
    }
    .resultContainer .btn{
        position:absolute;
        top:5px;
        right: 5px;
        width: 40px;
        height: 40px;
        font-size: 20px;
    }
    .btn{
        color: #fff;
        border: none;
        background-color: #3a3a3a;
        font-size: 16px;
        padding: 8px 12px;
        cursor: pointer;
    }
    .btn-large{
        display: block;
        width: 100%;
    }
    .setting{
        display: flex;
        justify-content: space-between;
        align-items: center;
        margin: 15px 0;
    }
}