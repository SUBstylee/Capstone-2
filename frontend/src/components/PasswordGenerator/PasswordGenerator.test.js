import { render } from '@testing-library/react';
import PasswordGenerator from './PasswordGenerator.jsx';

test("renders PasswordGenerator component without crashing", () => {
    const { getByText } = render(
        <PasswordGenerator />
    );

    expect(getByText('Password Generator')).toBeInTheDocument();
    expect(getByText('Password Length (6-20)')).toBeInTheDocument();
    expect(getByText('Include uppercase letters')).toBeInTheDocument();
    expect(getByText('Include lowercase letters')).toBeInTheDocument();
    expect(getByText('Include numbers')).toBeInTheDocument();
    expect(getByText('Password Length (6-20)')).toBeInTheDocument();
    expect(getByText('Include symbols')).toBeInTheDocument();
});

test("matches snapshot", function () {
    const { asFragment } = render(
        <PasswordGenerator />
    );
    expect(asFragment()).toMatchSnapshot();
});