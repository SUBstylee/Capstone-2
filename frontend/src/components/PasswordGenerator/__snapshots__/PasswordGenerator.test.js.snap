// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="PasswordGenerator"
  >
    <h2>
      Password Generator
    </h2>
    <div
      class="resultContainer"
    >
      <span
        id="result"
      />
      <button
        class="btn"
        id="clipboard"
      >
        <i
          class="far fa-clipboard"
        />
      </button>
    </div>
    <div
      class="settings"
    >
      <div
        class="setting"
      >
        <label
          for="length"
        >
          Password Length (6-20)
        </label>
        <input
          id="length"
          max="20"
          min="6"
          type="number"
          value="10"
        />
      </div>
      <div
        class="setting"
      >
        <label
          for="uppercase"
        >
          Include uppercase letters
        </label>
        <input
          checked=""
          id="uppercase"
          type="checkbox"
        />
      </div>
      <div
        class="setting"
      >
        <label
          for="lowercase"
        >
          Include lowercase letters
        </label>
        <input
          checked=""
          id="lowercase"
          type="checkbox"
        />
      </div>
      <div
        class="setting"
      >
        <label
          for="numbers"
        >
          Include numbers
        </label>
        <input
          checked=""
          id="numbers"
          type="checkbox"
        />
      </div>
      <div
        class="setting"
      >
        <label
          for="symbols"
        >
          Include symbols
        </label>
        <input
          checked=""
          id="symbols"
          type="checkbox"
        />
      </div>
    </div>
    <button
      class="btn btn-large"
      id="generate"
    >
      Generate Password
    </button>
  </div>
</DocumentFragment>
`;
